_posts/2019-07-09-javelin.md

```diff
@@ -0,0 +1,25 @@
+---
+layout: post
+title: Javelin project
+---
+
+
+We are having fun moving the gfx-rs project forward thanks to the power provided by the Rust language. We are building rich abstractions which expand beyond the gfx-hal API itself and into the internal layers of the backends, structured to be modular and maintainable. We are building high-performance graphics and compute infrastructure that deeply interacts with OS and drivers, thanks to the Rust's FFI capabilities and the lack of runtime. We do all of this while requiring only a portion of our developers' time and attention, who work on gfx-rs mostly as a side project. Rust allows us to move forward confidently and quickly, experiment with features as well as land the production code.
+
+But gfx-rs is not all written in Rust. There is one complex and important component we rely on that is written in a mixture of C and C++: [SPIRV-Cross](https://github.com/KhronosGroup/SPIRV-Cross). It's a shader translation library developed by [@TheMaister](https://github.com/HansKristian-Work) and a few Khronos members, although not an official Khronos product. We use it to generate platform-specific shaders from SPIR-V sources. It has a test suite, and its Metal backend is mostly developed and used by MoltenVK.
+
+SPIRV-Cross shows up quite a bit in our performance profiles (e.g. on Dota2). It is also written in a way (arguably!) that makes weak guarantees about what the state of the translator is at any point. And while it moves fast (in terms of contributions), it defines for us today what we can and what we can not do in terms of using advanced backend features, e.g. tessellation, argument buffers, etc. It complicates our [build process](https://github.com/gfx-rs/gfx/issues/2520), becoming a pain point for developers and users alike.
```

Comment:
Maybe we could rephrase
> It is also written in a way (arguably!) that makes weak guarantees about what the state of the translator is at any point.

to talk about how we'd prefer to work with immutable Rust data structures representing the compiler state directly instead of going through the Rust <-> C ABI <-> C++